lib/spack/llnl/util/filesystem.py

```diff
@@ -1117,6 +1117,18 @@ def visit_directory_tree(root, visitor, rel_path='', depth=0):
             visitor.after_visit_symlinked_dir(root, rel_child, depth)
 
 
+@system_path_filter
+def lightweight_traverse_tree(root, visit=None, *args, **kwargs):
```

Comment:
This function could benefit from fully implementing the visitor for `visit_directory_tree`:

* Avoid entering symlinks because we only need to visit each file once
* Do we need to update permissions on the symlinks themselves? (that isn't required on Linux)

Also regarding the name: the term `lightweight` is distracting and doesn't provide much info (e.g. one might think that this is the "lightweight" version of `traverse_tree` which it is not, and furthermore understanding what lightweight means requires reading both functions). I think `visit_recursive` would be more clear (and the docstring could say "this is like visit_directory_tree, but follows all symlinks).

Reply:
I'll have to investigate regarding the symlinks themselves, but I think if we're going to be implementing the Visitor, we may as well scrap this method (I thought we may want to do that), unless I'm misunderstanding you? 

Fully agreed on the name, I wasn't sure if this method would be dropped in preference of the actual traverse tree method.

Reply:
> we're going to be implementing the Visitor, we may as well scrap this method (I thought we may want to do that), unless I'm misunderstanding you?

To confirm: implementing the visitor seems like the best way to proceed IMO. I figure it still might make sense to have a separate method, unless `os.walk` can for example achieve the basic behavior of "visit each file in the tree, don't follow symlinks".